Fixes #1324: Properly closes connections in tribe

IRCody opened the pull request:
Fixes #1324

Summary of changes:
- adds proper http.Response.Close() calls. Primarily in worker.go (which causes this issue) but also around some of the client calls that I think could cause similar issues.

Testing done:
- Reproduced the bug in #1324 reliably, implemented the fix, and tested via same method and verified that the number of open connections isn't constantly growing. 

@intelsdi-x/snap-maintainers

thomastaylor312 commented on mgmt/tribe/worker/worker.go:
```diff
@@ -332,23 +332,27 @@ func (w worker) loadPlugin(plugin core.Plugin) error {
 			dir, err := ioutil.TempDir("", "")
 			if err != nil {
 				logger.Error(err)
+				resp.Body.Close()
```
Should this just be a `defer` right after the `c.TribeRequest()` call? It would get rid of the `Body.Close()` below

thomastaylor312 replied:
Yeah, makes sense.

candysmurf replied:
too many resp.Body.Close() in every error. Should it be added in the source?

lmroz replied:
So maybe move this part to separate function (and make this one less blobby), or just wrap it by ad hoc function
```
func() {
    ...
    resp := Do()
    iferr...
    defer resp.Body.Close()
    ...
}()
```

candysmurf commented:
LGTM

lmroz commented:
Go Documentation says
>The default HTTP client's Transport does not
>attempt to reuse HTTP/1.0 or HTTP/1.1 TCP connections
>("keep-alive") unless the Body is read to completion and is
>closed.
So it's good idea to read contents even if not used before closing.

lmroz commented:
Cannot add line comment to unchanged line, but here:
```
 func httpRespToAPIResp(rsp *http.Response) (*rbody.APIResponse, error) {
 	if rsp.StatusCode == 401 {
 		return nil, fmt.Errorf("Invalid credentials")  <<<<<<<<<<<<<<<<<
 	}
 	resp := new(rbody.APIResponse)
 	b, err := ioutil.ReadAll(rsp.Body)
rsp.Body.Close()
```
response body still has to be closed.

IRCody commented:
@lmroz, @candysmurf: I'm sure the mutliple rsp.Body.Close() calls can in all cases be reduced. When I was doing this fix I was focused on making it work. I can work on a version that is "cleaner" and refactors some of the ways we do things around http.Requests.

@lmroz: The code your linking already has the response body closed (I believe) because of the defer rsp.Body.Close() calls that happen prior to going into that function.

lmroz commented:
@IRCody Right, looks like body is closed just after this function is called. However this function is calling `.Close()` anyways, and that led me to believe that closing is responsibility of this function.  IMO it's good to decide which code is responsible for this. In case of calling `.Close()` twice current implementation just ignores it and doesn't even return error, however seems like it's not defined behavior.

IRCody commented:
@lmroz: I agree there is room for improvement. This code started out as debugging a live issue and iterating on fixing the issue. I need to go back and refactor so it's a cleaner fix. You're right that multiple calls to Close() are implementation defined.

IRCody commented:
@lmroz, @candysmurf, @thomastaylor312:

I added two commits that I think Improve this PR.  Before merging I'd like to merge the first two but I wanted to leave it as a separate commit for the moment so it's easy to see what changes have been made from the initial review.

The first commit addresses the comments in this PR around duplicate calls to Body.Close(). I refactored the downloading plugin part into another function that handles this so there is just one defer call to body.Close().

The second commit fixes an issue I noticed in the client where I think we are leaking http.Transports on client creation. The fix makes use of something that was added in go 1.7 (IdleConnTimeout in http.Transport).

Since this was added in 1.7 the 1.6.3 tests don't work. I talked with @nanliu about it we were wondering if it makes sense to peg Snap to more than the latest version of Go. Since the released binaries will always be the latest version. I was hoping to get some opinions on that here since it's relevant.

If we decide to continue doing latest-1 support I can update this PR to not have IdleConnTimeout and resubmit that change once go 1.8 is released.

/cc @intelsdi-x/snap-maintainers

nanliu commented:
We only build binaries with go 1.7.3. Since this is an application and not a library, we can safely drop 1.6.3.

geauxvirtual commented:
> We only build binaries with go 1.7.1. Since this is an application and not a library, we can safely drop 1.6.3. This will also reduce the test matrix significantly.

This makes the assumption that everyone using snap is using a release binary and is not building off of master.

IRCody commented:
>>We only build binaries with go 1.7.1. Since this is an application and not a library, we can safely drop 1.6.3. This will also reduce the test matrix significantly.

>This makes the assumption that everyone using snap is using a release binary and is not building off of master.

Or that they are comfortable running the latest version of Go. We already require relatively newer releases of go by only testing/supporting the 2 most recent releases.

lmroz commented on mgmt/tribe/worker/worker.go:
```diff
@@ -324,55 +324,72 @@ func (w worker) loadPlugin(plugin core.Plugin) error {
+		"_block":         "download-plugin",
+	})
+	resp, err := c.TribeRequest()
+	defer resp.Body.Close()
```
Nil pointer dereference possible here (on error `TribeRequest` will return nil response).

IRCody replied:
Good catch. I moved the defer below the err handling for tribeRequest.

lmroz commented on mgmt/tribe/worker/worker.go:
```diff
@@ -324,55 +324,72 @@ func (w worker) loadPlugin(plugin core.Plugin) error {
+			logger.Error(err)
+			return nil, err
+		}
+		f, err := os.Create(path.Join(dir, fmt.Sprintf("%s-%s-%d", plugin.TypeName(), plugin.Name(), plugin.Version())))
```
Not related to PR but `Create` is just `return OpenFile(name, O_RDWR|O_CREATE|O_TRUNC, 0666)`, so chmod can be done in more compact way.

IRCody replied:
Modified this so it's just one call to OpenFile.

lmroz commented on mgmt/tribe/worker/worker.go:
```diff
@@ -324,55 +324,72 @@ func (w worker) loadPlugin(plugin core.Plugin) error {
+		if resp.Header.Get("Content-Type") != "application/x-gzip" {
+			logger.WithField("content-type", resp.Header.Get("Content-Type")).Error("Expected application/x-gzip")
+		}
+		dir, err := ioutil.TempDir("", "")
```
But error during Create/Chmod (these are rather unlike) will leave that directory and just eats the path.

thomastaylor312 commented:
I agree with @nanliu and @IRCody on this one. Also, I would love to have this fix because it does cause issues with running tribe in Kubernetes that have the possibility of getting worse with scale

andrzej-k commented on mgmt/rest/client/client.go:
```diff
@@ -101,6 +102,17 @@ func Username(u string) metaOp {
 
+var (
+	secureTransport = &http.Transport{
+		TLSClientConfig: &tls.Config{InsecureSkipVerify: true},
```
```InsecureSkipVerify: true``` shouldn't this be set to ```false``` for secure transport?

IRCody replied:
Good catch @andrzej-k. I updated with the correct settings.

andrzej-k commented on mgmt/rest/client/client.go:
```diff
@@ -101,6 +102,17 @@ func Username(u string) metaOp {
+		IdleConnTimeout: time.Second,
+	}
+	insecureTransport = &http.Transport{
+		TLSClientConfig: &tls.Config{InsecureSkipVerify: false},
```
false -> true ?

IRCody commented:
I factored out the IdleConnTimeout for the transports so that this fix can be merged. I'll submit a separate PR with the ConnTimeout (that requires go1.7+) so discussion on that can happen separately.

cc: @candysmurf

candysmurf commented on mgmt/tribe/worker/worker.go:
```diff
@@ -324,55 +324,68 @@ func (w worker) loadPlugin(plugin core.Plugin) error {
+			return nil, err
+		}
+		io.Copy(f, resp.Body)
+		f.Close()
```
if using defer f.Close(), it'll take care of the error case and force close it anyway. will it?

candysmurf commented on mgmt/tribe/worker/worker.go:
```diff
@@ -324,55 +324,68 @@ func (w worker) loadPlugin(plugin core.Plugin) error {
+			logger.Error(err)
+			return err
+		}
+		if w.isPluginLoaded(plugin.Name(), plugin.TypeName(), plugin.Version()) {
```
add a comment. Making sure the plugin is loaded. Otherwise, I thought it should check this in the beginning before loading a plugin. Then I saw there is a similar check at the beginning. If no error is not good enough when loading a plugin?

IRCody replied:
I don't understand what you're saying.

candysmurf replied:
I meant to say:

   method: w.pluginManager.Load(rp) should give enough info of if a plugin is loaded or not. Why we need to check it again in line 343.  If line 343 is needed, please add a comment to it.

candysmurf commented:
LGTM
